### DISCUSSION TITLE
AMM for TEA: Uniswap vs Bancor (v3)

### DISCUSSION THREAD

**Discussion by geopor** [tearust/teaproject#131] [Polls]
Uniswap is the current automated market maker (AMM) that's being used in TEA Project's Epoch 12 -> https://app.uniswap.org/#/pool/45838

We can see that even though we provided most of the initial liquidity at a price around 1200 TEA / 1 Goerli ETH, market forces have made the price of TEA more valuable:

<img width="805" alt="Screenshot 2023-01-05 at 6 56 31 PM" src="https://user-images.githubusercontent.com/86096370/210921221-71b54da5-ffbe-43b6-b62c-97b23c389e84.png">

This brings us to the first possible drawback of using Uniswap v3, namely that liquidity must be provided within price ranges. It's only within these ranges that automated market making can happen. In the following graphic, liquidity is only provided along the curve from point A to B and stops outside that range.

<img width="516" alt="Screenshot 2023-01-05 at 3 37 19 PM" src="https://user-images.githubusercontent.com/86096370/210921604-9ee5afa2-0a78-40ef-9ae3-92268100d125.png">

A limitation with Uniswap v3 is when the price moves to the boundary of the price range, at which point the Uniswap LP sells all of one token for the other. The LP has ended at that point and is no longer able to provide liquidity until the price moves back in the liquidity range. There could be many such liquidity ranges for Uniswap LP pools, it all depends on the price ranges where the community is willing to provide liquidity.

And speaking of providing liquidity for Uniswap, there are a couple other drawbacks that depositors should be aware of:

- Impermanent loss. Anyone who has provided liquidity should be familiar with the loss in deposited value that occurs after the liquidity pool periodically rebalances itself and one token is sold for another.
- Forced token exposure - have to provide both sides of the liquidity pair.

## The Bancor v3 Alternative
Bancor is a competing AMM that's known for both its asymmetric pools and impermanent loss protection. V3 goes even further in limiting forced token exposure by using a single-sided liquidity model.

- User deposits one erc20 asset (currently around 150 accepted on the Bancor platform).
- User earns their share of every transaction based on their share of the pool.

A startup like the TEA Project just needs a market from which to trade our token, and saving 1/2 the investment required compared to a UNI or alternative is a very good deal for us.

## Bancor Depositors Don't Need to Stake

Every single-sided depositor is given an erc20 receipt token that doesn't need to be staked, it just naturally auto-compounds. Let's look at an example for how this works for someone who deposits BNT.

Users who deposit BNT to Bancor will also receive vBNT, Bancor’s governance token, in addition to bnBNT, the actual receipt token. Fees earned for providing liquidity are compounded back into the pool such that a user's bnBNT should grow over time. The following is an example of how a Bancor pool token would grow in value through the receipt token (bnBNT):

<img width="465" alt="Screenshot 2023-01-05 at 6 19 13 PM" src="https://user-images.githubusercontent.com/86096370/210923657-141831a6-fb3b-476a-81c5-bcaf2a5d1763.png">

## If Bancor Doesn't Use Pairs, How Does the AMM Work?
All trades in Bancor v3 are routed virtually through BNT: all tokens are virtually in one big shared liquidity pool known as the **Omnipool** with Bancor's BNT token on the other side of every trade. 

![IMG_4616](https://user-images.githubusercontent.com/86096370/210924342-8814bb7e-c8bd-47f3-811d-d5f7ad29dcdb.jpeg)

So for each trade, the price of the requested asset and the traded asset are given in terms of their value relative to BNT. 

## Virtual Liquidity Pools
Liquidity pools in Bancor3 are virtual as they do not contain any tokens. In effect, they are a logic layer above the vault that dictates the exchange of assets between the vault and a user.

![IMG_4615](https://user-images.githubusercontent.com/86096370/210925200-eb418642-e2cf-44d6-a228-feb8b311fdf0.jpeg)

Let's look at an example of a user selling **Token A** in order to get **Token B**. This is how it would look like in Bancor's virtual routing:

<img width="659" alt="Screenshot 2023-01-05 at 7 25 42 PM" src="https://user-images.githubusercontent.com/86096370/210924585-d9e7de0e-518f-45b7-ba9e-619d4c799ba7.png">

- The user sells **Token A** which is dumped into the Token A single-sided pool.
- Keeping in mind that both tokens were the same price (1 BNT) at the beginning of the transaction, 1 BNT is subtracted from Token A's virtual BNT ledger and moved into **Token B**'s virtual BNT ledger.
- The Token B that the user bought is now moved into their wallet and removed from the Token B pool.

## Surpluses and Deficits
As you can see from the above example, the Bancor v3 AMM creates fluctuating token supplies in pools as the market performance between trading pairs constantly creates deficits and surpluses. The token supply levels will always be a moving target and there will always be pools in deficit and some in surplus. This is simply because of the mechanics of the Bancor v3 AMM.

- When a token's market price dumps with respect to BNT, the pool becomes flooded with the token, creating a surplus in the pool. And conversely when the market price jumps for a token with respect to BNT, there will be a deficit in the pool.
- Every token position manages its own surplus / deficit situation. And whatever the deficit is, it'll remain constant until more trading happens on the pair.  
- Any project can elect to provide additional liquidity that serves as protection against deficit situations.

## How Would TEA Project Get Onto Bancor?
Any new project needs to be whitelisted by Bancor to get access to platform features like impermanent loss protection and auto-compounding. That said, new pool creation on Bancor is technically permissionless; anyone can acquire the necessary vBNT to create a proposal to onboard a new token. 

**When whitelisting is approved and the new project lists and provides their token for swaps, BNT is minted to match.**  This is considered the bootstrapping phase where deposits to the pool are allowed but no trading commences **until the exchange rate for the newly added token is specified by the BancorDAO.**

## Concerns About Adopting Bancor
To reiterate the consequences of Bancor's AMM mechanism: 

1. Each vault has a balance of tokens. 
2. The vault is in surplus when it contains more tokens than was directly supplied by liquidity providers. 
3. The vault is in deficit when it contains fewer tokens than liquidity providers deposited.

Currently all Bancor v3 vaults are in deficit status due to events that happened in 2022 (arbitrage attack together with bear market). You can see the deficit numbers at the bottom of their realtime dashboard: https://analytics.bancor.network

This has sapped value from the ecosystem and has had two major consequences:

1. User deposits have been slashed as their receipt tokens are worth less than they deposited.
2. It's no longer feasible to deposit funds into these pools that are in deficit.
If you deposited into them now, you'd experience something like what happened to the following user:

> I deposited some tokens into v3 and immediately lost half of them because it had an outstanding deficit.

Pools in deficit don't attract deposits as depositors will be at a deficit to start (entitled to less than they deposited.) This has understandably effectively halted deposits on Bancor3:

<img width="598" alt="Screenshot 2023-01-05 at 6 06 19 PM" src="https://user-images.githubusercontent.com/86096370/210927007-9aea77f6-619b-4517-8dbe-d33f01d1a641.png">

Although as a new pool we wouldn't face this issue, this is something to bear in mind.

**Comment by geopor** [comment]
A major concern is that if Bancor is our primary marketplace then we don’t have a “most liquid venue” for them to pull the starting price from. This could block us from getting whitelisted.

Another possible path is to use something like [Dodo's Crowdpooling](https://dodoex.github.io/docs/docs/crowdPooling/) which combines an IDO (which they **don't** promote) and rolls those funds into the initial liquidity pool.

**Comment by kevingzhang** [comment]
I am also interested in Dodo's crowdpooling idea https://dodoex.github.io/docs/docs/crowdPooling#why-crowdpooling

**Comment by geopor** [comment]
Kevin Zhang:
> Yes, this is exactly what Dodo offers. one stone two birds. IDO + Liquidity pool
> it is very important to have a relative large LP to start trading
> Most startup projects, like us, won’t have enough power to attract LP providers

DODO crowdpooling is permissionless, which means wee can launch it anytime. They actually don’t do any promotion for us, very similar to our relation to TApp tokens- they are adamant that they just provide the platform and don’t promote any of the projects that use it in order to limit their legal liability.

You can see a list of the projects that have used their crowdpooling here: https://community.dodoex.io/c/english/crowdpooling/25

Lots of small BNB chain projects (no raise was above $200k). They offer many other chains to launch on - ETH Polygon Arbitrum Optimism Avalanche - but liquidity would be tied to that chain. Looks like almost all of the projects using DODO launch on the BNB chain.

**Comment by geopor** [comment]
## Example of using DODO's crowd pool

Let's say a project uses the Dodo crowdpool to offer 1 million of their token **XYZ** at a $1 initial offering price.

Imagine there are 10 total bidders:
• Bidders 1 - 8 each want 100K at that price.
• Bidder 9 wants 200K at that price.
• Bidder 10 wants 1 million XYZ at that price.

Since the offering is oversubscribed by a factor of 2 ( $1 million worth of tokens, $2 million pledged), all the bidders above only get filled at 50% of their requested allotment. 

• Bidders 1-8 get 50K XYZ
• Bidder 9 100K XYZ
• Bidder 10 gets 500K XYZ

After the initial sale is completed, the **crowd** portion of Dodo's crowdpool is completed. Now we enter into the **pool** portion of the crowdpool:
the raised funds go into the liquidity pool, i.e. the 1 million USDT raised would go directly into the liquidity pool along with the other portion of the XYZ tokens set aside for the pool.

**Comment by kevingzhang** [comment]
I like the algorithm of Dodo, but I guess they do not have a lot of users yet.

**Reply by geopor** [reply]
Yes agree, and I notice many projects have adopted a similar algorithm that they deploy themselves. This is from a DeFi project on ETH (Antfarm):

> The Initial Liquidity Offering will take place place on January 17 2023 through a Smart Contract on Ethereum and will run for some days or until it reaches the hardcap (500 ETH). The ILO needs at least to reach the softcap (200 ETH) to be approved.
> During this sale 3.000.000 ATF tokens will be available (30% of the total supply) and the price will be fixed by the total contribution of the participants.
> All the funds raised through the Initial Sale will be used to deploy liquidity to ATF pairs. To do so, 3.000.000 other ATF tokens will be used  as well.

Especially since DODO won't do any promotion on our behalf it's technically not necessary to use their platform.

**Comment by kevingzhang** [comment]
I think I got it.
ATF did not Dodo paltform since Dodo cannot promote their project. But they did use Dodo's concept. They deployed their own smart contract to raise fund. The fund was locked in the smart contract, and was turned into the liquidity pool once the fund raising ends and swap starts. Do you know which swap platform are they using? If they did not use Dodo, they may use some other swap, such as Uniswap. Do you know? @geopor

**Reply by geopor** [reply]
Their liquidity positions are represented by NFTs so I suspect they just forked Uniswap v3.

**Reply by geopor** [reply]
Confirmed Uniswap v2 with elements of v3.

<img width="533" alt="Screenshot 2023-01-13 at 4 45 18 PM" src="https://user-images.githubusercontent.com/86096370/212442173-deff3c1d-d047-4814-9150-e063b56a363b.png">

**Comment by kevingzhang** [comment]
Given TEA is a stable (pegged to computing resources, not to any fiat currency), can we use TEA as a middle token like BNT?
So any other can be single-sided deposit while vTEA is provided as a counterpart? 

At least I found a few problems so far:
- TEA is not widely used yet. The oracle cannot get the exchange rate between any other token to TEA. So the PAMM won't work
- we will need to design the algorithm to make sure the vTEA can be always pegged to real TEA. We cannot mint TEA even on layer2.

There might be more questions. waiting for the answer. The point is , if we can, we can make a TEA Swap exchange
